### PR TITLE
Rewrite WaveTrack pasting as interval-major

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #5585

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

QA:  all the clip edge cases, mono and stereo...
- [x] Pasted-into track becomes selected
- [x] Cross-project paste, with and without project tempo differences
- [x] Paste with and without stretch ratio differences
- [x] With and without the new merging preference
- [x] Pasting into a clip, within one sample interval of time 0
- [x] With and without the "Edit clips can move" preference; test the prohibition
- [x] Moving of multiple other clips rightward to make room for the paste
- [x] New clips get correct names, and also when splitting to mono after the paste

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: libraries/lib-wave-track/WaveTrack.cpp
Comment (saintmatthieu) [inline libraries/lib-wave-track/WaveTrack.cpp:510]: Such progress is also used in `WaveTrack::Resample` in a way that doesn't seem correct to me (not taking the number of clips into account and the progression from one clip to the next). I was tempted to add a QA item for that but I expect that our testers won't be happy and yet this wouldn't be a regression.

**Review by saintmatthieu — APPROVED** [review]

File: libraries/lib-wave-track/WaveTrack.cpp
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:2520]: Correct.

I will remove more uses of `PasteOne` in two more PRs, keeping each PR not too large, not too slow to review, or to test.  Know that they are already in preparation.

Don't insist that I do all this sequentially.  I may want to have other PRs changing other parts of this big file also in flight concurrently through the project board.

Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:2520]: See #5602

**Comment by Paul-Licameli** [comment]
Rebased, after merging in the lastest hot-fix changes to WaveTrack.cpp, to be sure of no conflicts or need for additional changes -- there were none.

**Review by saintmatthieu — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
@saintmatthieu please review again.

I will squash it into fewer commits.

Importantly I had to figure out a silly mistake that put duplicate shared pointers to the same clip into the same track, fixed by one deleted line.

This PR is important to unblock other PRs dependent on it.  I hope it moves soon back to testing.

File: libraries/lib-wave-track/WaveTrack.cpp
Comment (saintmatthieu) [inline libraries/lib-wave-track/WaveTrack.cpp:2237]: I wonder why `GetIntervalAtTime` isn't sufficient.
In fact, looking at the `GetWideClipAtTime` implementation, I see logic that might have been relevant before the 2-of-6 PR, but which now seems futile to me. More about this in a follow-up comment closer to that implementation.

File: libraries/lib-wave-track/WaveTrack.cpp
Comment (saintmatthieu) [inline libraries/lib-wave-track/WaveTrack.cpp:3999]: Can this ever be executed?
Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).

**Review by saintmatthieu — COMMENTED** [review]
Review iteration complete.

File: libraries/lib-wave-track/WaveTrack.cpp
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:3999]: I want to make only a conservative refactoring that changes the iteration to interval major.

I did not spend my time reasoning through whether the old extra logic in `GetClipAtTime` is still important.  I assumed it was.

If you think this should simplify, and can prove it, and can also understand why @DavidBailes put that extra logic there, and can convince him too that it is no longer needed -- then I suggest you make a PR for that purpose.

But I believe that should not block approval of this more urgent cleanup.

Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:3999]: > Can this ever be executed? Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).

Did you notice that the iteration through the clips was in backward order?

Comment (saintmatthieu) [reply libraries/lib-wave-track/WaveTrack.cpp:3999]: https://github.com/audacity/audacity/pull/5586#discussion_r1404086453 intended to prove that already, and to my sense did. From what I read you haven't examined if that extra old logic was useful or not, and yet you introduced a new method replicating this. If `GetIntervalAtTime` didn't exist before I wouldn't mind so much, but it does, and introducing an alternative and even replacing existing usages suggests there's something wrong with it.

Comment (saintmatthieu) [reply libraries/lib-wave-track/WaveTrack.cpp:3999]: > Did you notice that the iteration through the clips was in backward order?

I don't see how this impacts the reasoning in https://github.com/audacity/audacity/pull/5586#discussion_r1404086453.

Comment (DavidBailes) [reply libraries/lib-wave-track/WaveTrack.cpp:3999]: > Can this ever be executed? Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).

@saintmatthieu 
The last time I changed this this function the initial lines were:
```
const auto clips = SortedClipArray();
   auto p = std::find_if(clips.rbegin(), clips.rend(), [&] (WaveClip* const& clip) {
      return time >= clip->GetStartTime() && time <= clip->GetEndTime(); });
```
Since then, it has been changed so that if a time is equal to the time of the end of the clip it is not considered to be in the clip.
So as you say, the lines that correct for rounding errors are no longer necessary.

Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:3999]: > > Can this ever be executed? Say `clipB->WithinPlayRegion(t)` returns true, it then becomes `p`. Then further down you test that `t == p->GetPlayEndTime()`, which can never be true, since `p->WithinPlayRegion(t) == true` implies that `t < p->GetPlayEndTime()` (or there is a bug).
> 
> @saintmatthieu The last time I changed this this function the initial lines were:
> 
> ```
> const auto clips = SortedClipArray();
>    auto p = std::find_if(clips.rbegin(), clips.rend(), [&] (WaveClip* const& clip) {
>       return time >= clip->GetStartTime() && time <= clip->GetEndTime(); });
> ```
> 
> Since then, it has been changed so that if a time is equal to the time of the end of the clip it is not considered to be in the clip. So as you say, the lines that correct for rounding errors are no longer necessary. (And the times of the start and end of clips are now rounded to multiples of samples.)

You made changes several years ago, and I don't remember the motivating use case to test.  Do you?  Was it clip movement with the keys?

Comment (DavidBailes) [reply libraries/lib-wave-track/WaveTrack.cpp]: Yes, it was clip movement using keystrokes. It was to ensure if a user moved to the end of a clip which was immediately adjacent to a following clip, that the keystrokes to move clips would move the second clip and not the first. The definition of  WaveClip::WithinPlayRegion() ensures that this will happen.

**Comment by chinakov** [comment]
Found one issue so far: when a User tries to paste a smart clip, either on:

- same track, 
- new track
- cross-project with the same or different tempo

the smart clip no longer contains the hidden data. 

This issue persists on both mono and stereo tracks.

**Review by saintmatthieu — APPROVED** [review]
No bug detected but zero to two simplifications.

**Comment by chinakov** [comment]
Tested, all issues fixed. Thanks!

**Pull request merged**
